Multiplexes messages sent to ModelProcess from WebProcess

eddydas opened the pull request:
#### a990f17d6abe0458febd3c8b58a4daf66bfa9871
<pre>
Multiplexes messages sent to ModelProcess from WebProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=269631">https://bugs.webkit.org/show_bug.cgi?id=269631</a>
<a href="https://rdar.apple.com/123122822">rdar://123122822</a> (Multiplexing IPC messages)

Reviewed by Tim Horton.

Implements IPC messages multiplexing for Model process to handle
distinct model tags.

On WebProcess side, added:
ModelProcessModelPlayerManager, a per-WebProcess singleton, which creates/manages all
ModelProcessModelPlayer for the same WebProcess.

On ModelProcess side, added:
ModelProcessModelPlayerManagerProxy, which exists 1:1 to ModelConnectionToWebProcess,
ModelProcessModelPlayerManagerProxy creates/manages all
ModelProcessModelPlayerProxy for the same WebProcess.

Added IPC dispatching mechanism so that ModelProcessModelPlayer
can easily send a message to its corresponding ModelProcessModelPlayerProxy,
and vice versa.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ModelPlayerIdentifier.h:
Added. This identifies different model tags from the same WebProcess.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
(WebKit::ModelConnectionToWebProcess::dispatchMessage):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
(WebKit::ModelConnectionToWebProcess::protectedConnection):
(WebKit::ModelConnectionToWebProcess::modelProcessModelPlayerManagerProxy):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp: Added.
(WebKit::ModelProcessModelPlayerManagerProxy::ModelProcessModelPlayerManagerProxy):
(WebKit::ModelProcessModelPlayerManagerProxy::~ModelProcessModelPlayerManagerProxy):
(WebKit::ModelProcessModelPlayerManagerProxy::clear):
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
(WebKit::ModelProcessModelPlayerManagerProxy::deleteModelPlayer):
(WebKit::ModelProcessModelPlayerManagerProxy::didReceivePlayerMessage):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.h: Added.
(WebKit::ModelProcessModelPlayerManagerProxy::modelConnectionToWebProcess):
(WebKit::ModelProcessModelPlayerManagerProxy::didReceiveMessageFromWebProcess):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.cpp: Copied from Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp.
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::invalidate):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
An example stub message to be fleshed out later.

* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.h: Copied from Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp.
* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.messages.in: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
(WebKit::ModelProcessConnection::didClose):
(WebKit::ModelProcessConnection::dispatchMessage):
(WebKit::ModelProcessConnection::createVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::destroyVisibilityPropagationContextForPage):
Corrected typo (from Copy and Paste) from before. This has nothing to do with GPU.

(WebKit::ModelProcessConnection::configureLoggingChannel):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
(WebKit::ModelProcessConnection::addClient):
(WebKit::ModelProcessConnection::removeClient): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp: Renamed from Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.mm.
This file can be .cpp instead of .mm

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
(WebKit::ModelProcessModelPlayer::identifier):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in: Copied from Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp: Added.
(WebKit::ModelProcessModelPlayerManager::create):
(WebKit::ModelProcessModelPlayerManager::modelProcessConnection):
(WebKit::ModelProcessModelPlayerManager::createModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::deleteModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::didReceivePlayerMessage):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h: Added.
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::createModelPlayer):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::modelProcessModelPlayerManager):
WebProcess now owns a singleton of ModelProcessModelPlayerManager.

Canonical link: <a href="https://commits.webkit.org/275140@main">https://commits.webkit.org/275140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d54773c5a6294f26a5826aa91890852fd588f984

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40972 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43350 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43531 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37064 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17316 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33940 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16911 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35310 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14565 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/40842 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14678 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36303 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44832 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37165 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36615 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12957 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38725 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17406 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9200 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17457 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -80,13 +80,13 @@
+static_assert(std::is_same_v<WTF::ProcessID, int>);
 #endif
 #if !(OS(WINDOWS))
-static_asssert(std::is_same_v<WTF::ProcessID, pid_t>);
```
:+1:

hortont424 replied:
how??

eddydas replied:
I changed messages.py, and it asks me to:
```
# When updating this list, run "make -C Source/WebKit/Scripts/webkit/tests" to keep the webkitpy tests passing.
```
and this diff is what it produces.

djg reviewed (commented):
I did a lite review but someone with more experience of IPC should review.

hortont424 commented on Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
```diff
@@ -40,11 +40,14 @@
 #include "IPCTester.h"
 #endif
 
+#include <WebCore/NowPlayingManager.h>
```
What is this doing here??

eddydas replied:
My mistake! Thanks for catching that.

hortont424 commented on Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
```diff
@@ -96,12 +98,14 @@ void ModelProcessConnection::invalidate()
 
 void ModelProcessConnection::didClose(IPC::Connection&)
 {
+    RELEASE_LOG_ERROR(Process, "%p - ModelProcessConnection::didClose", this);
```
Is this really an error?

eddydas replied:
Not really! Lemme change it to normal log.

hortont424 commented on Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
```diff
@@ -143,18 +152,18 @@ bool ModelProcessConnection::waitForDidInitialize()
 
 void ModelProcessConnection::configureLoggingChannel(const String& channelName, WTFLogChannelState state, WTFLogLevel level)
 {
-    connection().send(Messages::GPUConnectionToWebProcess::ConfigureLoggingChannel(channelName, state, level), { });
```
Whoops :)

eddydas committed:
Multiplexes messages sent to ModelProcess from WebProcess
https://bugs.webkit.org/show_bug.cgi?id=269631
rdar://123122822 (Multiplexing IPC messages)

Reviewed by Tim Horton.

Implements IPC messages multiplexing for Model process to handle
distinct model tags.

On WebProcess side, added:
ModelProcessModelPlayerManager, a per-WebProcess singleton, which creates/manages all
ModelProcessModelPlayer for the same WebProcess.

On ModelProcess side, added:
ModelProcessModelPlayerManagerProxy, which exists 1:1 to ModelConnectionToWebProcess,
ModelProcessModelPlayerManagerProxy creates/manages all
ModelProcessModelPlayerProxy for the same WebProcess.

Added IPC dispatching mechanism so that ModelProcessModelPlayer
can easily send a message to its corresponding ModelProcessModelPlayerProxy,
and vice versa.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ModelPlayerIdentifier.h:
Added. This identifies different model tags from the same WebProcess.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.cpp:
(WebKit::ModelConnectionToWebProcess::ModelConnectionToWebProcess):
(WebKit::ModelConnectionToWebProcess::dispatchMessage):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.h:
(WebKit::ModelConnectionToWebProcess::protectedConnection):
(WebKit::ModelConnectionToWebProcess::modelProcessModelPlayerManagerProxy):
* Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in:
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.cpp: Added.
(WebKit::ModelProcessModelPlayerManagerProxy::ModelProcessModelPlayerManagerProxy):
(WebKit::ModelProcessModelPlayerManagerProxy::~ModelProcessModelPlayerManagerProxy):
(WebKit::ModelProcessModelPlayerManagerProxy::clear):
(WebKit::ModelProcessModelPlayerManagerProxy::createModelPlayer):
(WebKit::ModelProcessModelPlayerManagerProxy::deleteModelPlayer):
(WebKit::ModelProcessModelPlayerManagerProxy::didReceivePlayerMessage):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.h: Added.
(WebKit::ModelProcessModelPlayerManagerProxy::modelConnectionToWebProcess):
(WebKit::ModelProcessModelPlayerManagerProxy::didReceiveMessageFromWebProcess):
* Source/WebKit/ModelProcess/ModelProcessModelPlayerManagerProxy.messages.in: Added.
* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.cpp: Copied from Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp.
(WebKit::ModelProcessModelPlayerProxy::create):
(WebKit::ModelProcessModelPlayerProxy::ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::~ModelProcessModelPlayerProxy):
(WebKit::ModelProcessModelPlayerProxy::invalidate):
(WebKit::ModelProcessModelPlayerProxy::loadModel):
An example stub message to be fleshed out later.

* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.h: Copied from Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp.
* Source/WebKit/ModelProcess/ModelProcessModelPlayerProxy.messages.in: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.cpp:
(WebKit::ModelProcessConnection::didClose):
(WebKit::ModelProcessConnection::dispatchMessage):
(WebKit::ModelProcessConnection::createVisibilityPropagationContextForPage):
(WebKit::ModelProcessConnection::destroyVisibilityPropagationContextForPage):
Corrected typo (from Copy and Paste) from before. This has nothing to do with GPU.

(WebKit::ModelProcessConnection::configureLoggingChannel):
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
(WebKit::ModelProcessConnection::addClient):
(WebKit::ModelProcessConnection::removeClient): Deleted.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp: Renamed from Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.mm.
This file can be .cpp instead of .mm

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
(WebKit::ModelProcessModelPlayer::identifier):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.messages.in: Copied from Source/WebKit/ModelProcess/ModelConnectionToWebProcess.messages.in.
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp: Added.
(WebKit::ModelProcessModelPlayerManager::create):
(WebKit::ModelProcessModelPlayerManager::modelProcessConnection):
(WebKit::ModelProcessModelPlayerManager::createModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::deleteModelProcessModelPlayer):
(WebKit::ModelProcessModelPlayerManager::didReceivePlayerMessage):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h: Added.
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::createModelPlayer):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::WebProcess):
* Source/WebKit/WebProcess/WebProcess.h:
(WebKit::WebProcess::modelProcessModelPlayerManager):
WebProcess now owns a singleton of ModelProcessModelPlayerManager.

Canonical link: https://commits.webkit.org/275140@main
